From 3e428c616f0201fad25ce7ffb62a9e624cd33765 Mon Sep 17 00:00:00 2001
From: Lynne <dev@lynne.ee>
Date: Thu, 13 Mar 2025 16:26:40 +0000
Subject: [PATCH 01/14] BACKPORT: FFHWAccel: add buffer_ref argument to
 start_frame

This commit adds a reference to the buffer as an argument to
start_frame, and adapts all existing code.

This allows for asynchronous hardware accelerators to skip
copying packet data by referencing it.
(cherry picked from commit caff29dbb18feeb87cb00fc4c33d20cf01667be0)

Only include start_frame buffer and size parameter change in backport.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 libavcodec/proresdec.c | 2 +-
 libavcodec/vp9.c       | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/libavcodec/proresdec.c b/libavcodec/proresdec.c
index 6a256107b494d..990f918b35dbb 100644
--- a/libavcodec/proresdec.c
+++ b/libavcodec/proresdec.c
@@ -805,7 +805,7 @@ static int decode_frame(AVCodecContext *avctx, AVFrame *frame,
 
     if (avctx->hwaccel) {
         const FFHWAccel *hwaccel = ffhwaccel(avctx->hwaccel);
-        ret = hwaccel->start_frame(avctx, NULL, 0);
+        ret = hwaccel->start_frame(avctx, avpkt->data, avpkt->size);
         if (ret < 0)
             return ret;
         ret = hwaccel->decode_slice(avctx, avpkt->data, avpkt->size);
diff --git a/libavcodec/vp9.c b/libavcodec/vp9.c
index 8ede2e2eb3587..246eafb5f4e2e 100644
--- a/libavcodec/vp9.c
+++ b/libavcodec/vp9.c
@@ -1610,7 +1610,7 @@ static int vp9_decode_frame(AVCodecContext *avctx, AVFrame *frame,
 
     if (avctx->hwaccel) {
         const FFHWAccel *hwaccel = ffhwaccel(avctx->hwaccel);
-        ret = hwaccel->start_frame(avctx, NULL, 0);
+        ret = hwaccel->start_frame(avctx, pkt->data, pkt->size);
         if (ret < 0)
             return ret;
         ret = hwaccel->decode_slice(avctx, pkt->data, pkt->size);

